### ISSUE TITLE
Cleanup Error

### ISSUE THREAD

**Issue by thehydroimpulse** [ember-cli/ember-cli#3486]
I'm getting this error on CircleCI but not locally:

```
Cleanup error.
Object function glob(pattern, options, cb) {
  if (typeof options === "function") cb = options, options = {}
  if (!options) options = {}

  if (typeof options === "number") {
    deprecated()
    return
  }

  var g = new Glob(pattern, options, cb)
  return g.sync ? g.found : g
} has no method 'hasMagic'
    at Function.rimrafSync [as sync] (/home/ubuntu/stacks/client/node_modules/ember-cli/node_modules/broccoli-caching-writer/node_modules/rimraf/rimraf.js:262:13)
    at ConfigLoader.CachingWriter.cleanup (/home/ubuntu/stacks/client/node_modules/ember-cli/node_modules/broccoli-caching-writer/index.js:114:10)
    at cleanupTree (/home/ubuntu/stacks/client/node_modules/ember-cli/node_modules/broccoli/lib/builder.js:128:19)
    at /home/ubuntu/stacks/client/node_modules/ember-cli/node_modules/broccoli/node_modules/promise-map-series/index.js:11:14
    at $$$internal$$tryCatch (/home/ubuntu/stacks/client/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:490:16)
    at $$$internal$$invokeCallback (/home/ubuntu/stacks/client/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:502:17)
    at $$$internal$$publish (/home/ubuntu/stacks/client/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:473:11)
    at Object.$$rsvp$asap$$flush [as _onImmediate] (/home/ubuntu/stacks/client/node_modules/ember-cli/node_modules/rsvp/dist/rsvp.js:1581:9)
    at processImmediate [as _immediateCallback] (timers.js:354:15)
```

I'm running version 0.1.12.

**Comment by avdv** [comment]
I stumbled over the same problem. I'm using Jenkins, ember-cli version 0.1.15.

Running `npm install` as part of my Jenkins build, spew out these warnings:

```
npm WARN unmet dependency /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-caching-writer/node_modules/rimraf requires glob@'^4.4.2' but will load
npm WARN unmet dependency /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/glob,
npm WARN unmet dependency which is version 4.0.5
npm WARN unmet dependency /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-es6modules/node_modules/broccoli-caching-writer/node_modules/rimraf requires glob@'^4.4.2' but will load
npm WARN unmet dependency /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/glob,
npm WARN unmet dependency which is version 4.0.5
npm WARN unmet dependency /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-es6modules/node_modules/esperanto/node_modules/sander/node_modules/rimraf requires glob@'^4.4.2' but will load
npm WARN unmet dependency /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/glob,
npm WARN unmet dependency which is version 4.0.5
npm WARN unmet dependency /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-sourcemap-concat/node_modules/broccoli-caching-writer/node_modules/rimraf requires glob@'^4.4.2' but will load
npm WARN unmet dependency /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/glob,
npm WARN unmet dependency which is version 4.0.5
```
Whereas in my local dev folder, all was well.

When trying to run `ember test` I received the following error and the build never finishes:

```
Cleanup error.
undefined is not a function
TypeError: undefined is not a function
    at Function.rimrafSync (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-caching-writer/node_modules/rimraf/rimraf.js:262:13)
    at ConfigLoader.CachingWriter.cleanup (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-caching-writer/index.js:114:10)
    at cleanupTree (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli/lib/builder.js:128:19)
    at /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/index.js:11:14
    at lib$rsvp$$internal$$tryCatch (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/node_modules/rsvp/dist/rsvp.js:489:16)
    at lib$rsvp$$internal$$invokeCallback (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/node_modules/rsvp/dist/rsvp.js:501:17)
    at lib$rsvp$$internal$$publish (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/node_modules/rsvp/dist/rsvp.js:472:11)
    at lib$rsvp$asap$$flush (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/node_modules/rsvp/dist/rsvp.js:1290:9)
    at process._tickCallback (node.js:355:11)
Build failed.
undefined is not a function
TypeError: undefined is not a function
    at rimraf (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-caching-writer/node_modules/rimraf/rimraf.js:57:13)
    at lib$rsvp$node$$tryApply (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-caching-writer/node_modules/rsvp/dist/rsvp.js:1467:11)
    at lib$rsvp$node$$handleValueInput (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-caching-writer/node_modules/rsvp/dist/rsvp.js:1567:20)
    at fn (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-caching-writer/node_modules/rsvp/dist/rsvp.js:1555:18)
    at /tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/broccoli-caching-writer/index.js:100:14
    at lib$rsvp$$internal$$tryCatch (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/node_modules/rsvp/dist/rsvp.js:489:16)
    at lib$rsvp$$internal$$invokeCallback (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/node_modules/rsvp/dist/rsvp.js:501:17)
    at lib$rsvp$$internal$$publish (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/node_modules/rsvp/dist/rsvp.js:472:11)
    at lib$rsvp$asap$$flush (/tmp/JENHOME/jenkins/jobs/adam-build/workspace/node_modules/ember-cli/node_modules/promise-map-series/node_modules/rsvp/dist/rsvp.js:1290:9)
    at process._tickCallback (node.js:355:11)
```

The reason for the error seems to be that an old version of the [node-glob](https://github.com/isaacs/node-glob) library is used. Looking at its code, version 4.2.0 was the first release which contained the `hasMagic` method.

Double checking the versions confirmed that in my dev folder, npm had pulled down version 4.4.0 whereas the Jenkins workspace only had version 4.0.5.

I guess `ember new` generated this line in package.json when we kicked of this project:

```
    "glob": "^4.0.5",
```

Apparently, there's a difference between calling `ember install` and `npm install && bower install`.  I reverted to use the latter because `ember install` always bailed out early when a bower or npm dependency was missing if there already was a bower_components or node_modules directory, respectively:

```
% ember install
version: 0.1.15

Missing npm packages: 
Package: glob
  * Specified: ^4.0.5
  * Installed: (not installed)

Run `npm install` to install missing dependencies.
```

## TL;DR

Now, in my build I clean up the directory before doing a build, ie. "git clean -fdx" and use "ember install" instead.

**Comment by regularjack** [comment]
To fix this issue add `"rimraf": "2.2.8"` to your package.json. See https://github.com/ember-cli/ember-cli/issues/3413#issuecomment-77104018 for more details.

**Comment by rwjblue** [comment]
Ember CLI 0.2.0 locks rimraf to 2.2.8 and glob to 4.0.5 to deal with these issues.  We have a path forward to update glob + rimraf, and I expect to be able to push that out in a future release.

If you can upgrade to 0.2.0, it should resolve the issue. Otherwise, you can lock your project to `rimraf@2.2.8` and `glob@4.0.5`.

**Comment by alexBaizeau** [comment]
I started using [npm shrinkwrap](https://docs.npmjs.com/cli/shrinkwrap) it also solves the problem going forward if you generate your `npm-shrinkwrap.json` file from a good working project.

**Comment by thehydroimpulse** [comment]
Thanks @avdv @rwjblue, upgrading fixed the issue.

**Comment by srsgores** [comment]
I just got this, using ``ember-cli 0.2.0-beta.1``.  Not sure if it's related:

```
Cleanup error.
undefined is not a function
TypeError: undefined is not a function
    at Function.rimrafSync (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\broccoli-caching-writer\node_modu
les\rimraf\rimraf.js:262:13)
    at ConfigLoader.CachingWriter.cleanup (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\broccoli-caching-w
riter\index.js:114:10)
    at cleanupTree (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\broccoli\lib\builder.js:128:19)
    at C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series\index.js:11:14
    at lib$rsvp$$internal$$tryCatch (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series\node_
modules\rsvp\dist\rsvp.js:489:16)
    at lib$rsvp$$internal$$invokeCallback (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series
\node_modules\rsvp\dist\rsvp.js:501:17)
    at lib$rsvp$$internal$$publish (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series\node_m
odules\rsvp\dist\rsvp.js:472:11)
    at lib$rsvp$asap$$flush (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series\node_modules\
rsvp\dist\rsvp.js:1290:9)
    at process._tickCallback (node.js:355:11)
Build failed.
undefined is not a function
TypeError: undefined is not a function
    at rimraf (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\broccoli-caching-writer\node_modules\rimraf\ri
mraf.js:57:13)
    at lib$rsvp$node$$tryApply (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\broccoli-caching-writer\node_
modules\rsvp\dist\rsvp.js:1467:11)
    at lib$rsvp$node$$handleValueInput (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\broccoli-caching-writ
er\node_modules\rsvp\dist\rsvp.js:1567:20)
    at fn (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\broccoli-caching-writer\node_modules\rsvp\dist\rsv
p.js:1555:18)
    at C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\broccoli-caching-writer\index.js:100:14
    at lib$rsvp$$internal$$tryCatch (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series\node_
modules\rsvp\dist\rsvp.js:489:16)
    at lib$rsvp$$internal$$invokeCallback (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series
\node_modules\rsvp\dist\rsvp.js:501:17)
    at lib$rsvp$$internal$$publish (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series\node_m
odules\rsvp\dist\rsvp.js:472:11)
    at lib$rsvp$asap$$flush (C:\Users\_admin\Documents\GitHub\ember-accessible-icon\node_modules\ember-cli\node_modules\promise-map-series\node_modules\
rsvp\dist\rsvp.js:1290:9)
    at process._tickCallback (node.js:355:11)

1..0
# tests 0
# pass  0
# fail  0

# ok
No tests were run, please check whether any errors occurred in the page (ember test --server) and ensure that you have a test launcher (e.g. PhantomJS)
enabled.
```

node ``0.12.0`` x64, npm ``2.5.1``

**Comment by Truffula** [comment]
`hasMagic` seems to have been added in glob `v4.2.0` and is used . @srsgores I updated to `v4.5.3` (with `npm update glob`) and it now builds.
